### PR TITLE
Refactoring for configurable DNS upstreams & some other refinements

### PR DESCRIPTION
**Pull request by Mygod**
I am not going to pretend I know what I am doing with these code.

Elon Musk

### CODE REVIEW

File: src/relay/dnsrelay/mod.rs [hunk 1]
```diff
@@ -225,8 +143,22 @@ pub async fn run(context: SharedContext) -> io::Result<()> {
...
+    #[cfg(target_os = "android")]
+    let local_upstream = Arc::new(upstream::UnixSocketUpstream {
+        path: config.local_dns_path.clone().expect("local query DNS path"),
+    });
+    #[cfg(not(target_os = "android"))]
+    let local_upstream = Arc::new(upstream::UdpUpstream {
+        server: config.local_dns_addr.clone().expect("local query DNS address"),
+    });
     // FIXME: We use TCP to send remote queries by default, which should be configuable.
     let balancer = PlainPingBalancer::new(context.clone(), ServerType::Tcp).await;
```
Comment (Mygod) [inline src/relay/dnsrelay/mod.rs:156]: @zonyitoo Shouldn't we use `Arc::clone` for `SharedContext`?

**Comment by Mygod** [comment]
@madeye @zonyitoo Please review it more carefully I really do not know what I am doing.

**Comment by madeye** [comment]
Can we also have this? https://github.com/madeye/shadowsocks-rust/commit/055b4f63be3571de43cb3010c364e61a1546cad7

**Comment by Mygod** [comment]
With these changes, Android client should behave reasonably with any reasonable ACL file.

File: src/relay/dnsrelay/mod.rs [hunk 2]
```diff
@@ -1,194 +1,90 @@
...
             RData::A(ref ip) => context.check_ip_in_proxy_list(&IpAddr::from(*ip)),
             RData::AAAA(ref ip) => context.check_ip_in_proxy_list(&IpAddr::from(*ip)),
-            _ => true,
+            RData::PTR(_) => panic!("PTR records should not reach here"),
+            _ => context.is_default_in_proxy_list(),
         };
         if !forward {
-            debug!("pick local response (ip): {:?}", local_response);
-            return Ok((local_response, false));
+            debug!("pick local response (response): {:?}", local_response);
```
Comment (madeye) [inline src/relay/dnsrelay/mod.rs:80]: Value borrowed here after move here:

https://github.com/shadowsocks/shadowsocks-rust/pull/242/files#diff-0fe08b50f6647bbea91fb44106ea4d16R68

File: src/relay/dnsrelay/mod.rs [hunk 3]
```diff
@@ -1,194 +1,90 @@
...
-        let remote_response = remote_response_fut.await.unwrap_or_else(Message::new);
-        return Ok((remote_response, true));
-    }
-
+    // FIXME: spawn(remote_response_fut)
+    let local_response = local_response_fut.await.unwrap_or_else(|_| Message::new());
     for rec in local_response.answers() {
+        if rec.record_type() != query.query_type() {
+            warn!("local DNS response has inconsistent answer type {} for query {}", rec.record_type(), query);
+            break
```
Comment (madeye) [inline src/relay/dnsrelay/mod.rs:71]: This line looks problematic, it's possible we get `A` answers for  `CNAME` queries.

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:71]: A typical test case is `www.baidu.com`. Local DNS server returns `A` records even if query is `CNAME`, leading to wrong ACL routing, as we're assuming any domain name resolved by remote should be proxied.

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:71]: Okay, it's kind of DNS hijacking issue of my local ISP DNS server, but we'd better refine the ACL part to not proxy any hostname resolved by remote.

Comment (madeye) [reply src/relay/dnsrelay/mod.rs:71]: Here is my fix: https://github.com/madeye/shadowsocks-rust/commit/98057a2e3ddfadffe57a762075759bec4ca1e8db

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:71]:
```
$ dig @8.8.8.8 -t CNAME baidu.com

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> @8.8.8.8 -t CNAME baidu.com
; (1 server found)
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 37016
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;baidu.com.                     IN      CNAME

;; AUTHORITY SECTION:
baidu.com.              1799    IN      SOA     dns.baidu.com. sa.baidu.com. 2012142433 300 300 2592000 7200
```

To quote rfc1034:

> Answer          Carries RRs which _directly_ answer the query.
> ...
> Additional      Carries RRs which may be helpful in using the RRs in the
>                 other sections.

Under normal circumstances, only `A`/`AAAA`/`PTR` queries are issued.

P.S. An unpolluted DNS exchange with additional sections looks like --

```
$ dig -t MX google.com

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> -t MX google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 57293
;; flags: qr rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 7

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;google.com.                    IN      MX

;; ANSWER SECTION:
google.com.             600     IN      MX      50 alt4.aspmx.l.google.com.
google.com.             600     IN      MX      40 alt3.aspmx.l.google.com.
google.com.             600     IN      MX      30 alt2.aspmx.l.google.com.
google.com.             600     IN      MX      10 aspmx.l.google.com.
google.com.             600     IN      MX      20 alt1.aspmx.l.google.com.

;; ADDITIONAL SECTION:
alt2.aspmx.l.google.com. 152    IN      A       172.253.116.26
aspmx.l.google.com.     34      IN      A       172.217.197.26
alt1.aspmx.l.google.com. 177    IN      A       64.233.186.27
alt1.aspmx.l.google.com. 39     IN      AAAA    2800:3f0:4003:c00::1b
alt4.aspmx.l.google.com. 280    IN      A       172.217.218.26
alt3.aspmx.l.google.com. 280    IN      A       172.253.120.26
```

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:71]: tl;dr I believe this is fine.

File: src/relay/dnsrelay/mod.rs (hunk 2 above)
Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:80]: I didn't get a compile error though.

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:80]: Okay I see it now.

File: src/relay/dnsrelay/mod.rs (hunk 3 above)
Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:71]: Okay rfc1034 does permit additional CNAME answers, which is a different issue.

```
   - The answer to the query, possibly preface by one or more CNAME
     RRs that specify aliases encountered on the way to an answer.
```

File: src/relay/dnsrelay/mod.rs (hunk 1 above)
Comment (zonyitoo) [reply src/relay/dnsrelay/mod.rs:134]: What do you mean? `context.clone()` calls `SharedContext::clone`, which is `Arc::clone`. No?

Comment (Mygod) [reply src/relay/dnsrelay/mod.rs:134]: I see you are right. https://doc.rust-lang.org/std/sync/struct.Arc.html#cloning-references

**Comment by madeye** [comment]
49966d9 seems breaking everything...

**Comment by Mygod** [comment]
@madeye Are you seeing `'async fn' resumed after completion`? Or something else?

**Comment by Mygod** [comment]
Try 13bcc5d09595aaa46d581438eebda0c5d422ad0d.

**Comment by Mygod** [comment]
@madeye @zonyitoo Done for today. Want to review these first?

File: src/relay/dnsrelay/mod.rs (hunk 3 above)
Comment (Mygod) [reply src/relay/dnsrelay/mod.rs]: Resolved via cb62eff.

**Comment by madeye** [comment]
`13bcc5d` works, but then `9712a97` breaks ACL again.

To reproduce the issue, please use bypass China route.

**Comment by madeye** [comment]
d1a3ac6 should fix the issue.

**Comment by zonyitoo** [comment]
I am getting more and more confuse about what these code are working for. LoL.

Nevermind. Since @madeye has approved, then please @Mygod merge master and fix conflicts.

**Comment by Mygod** [comment]
@zonyitoo In that case feel free to comment on where the code doesn't make sense and I can try to explain... Also maybe just merge this instead of rebasing it? There are too many commits now and I don't want to squash everything.

**Comment by zonyitoo** [comment]
I saw in ACL, resolving `DomainName` to IPs and then check whether IPs are in configured lists, have been removed. Why?

**Comment by zonyitoo** [comment]
Ah, my fault. I found that now.

**Pull request merged**